lib/World.luau

```diff
@@ -117,8 +205,278 @@ end
 	@return number
 	@return {[Component]: ComponentInstance}
 ]=]
-function World:__iter()
-	return World._next, self
+function World.__iter(world: World)
+	local lastEntityId = nil
+	return function(): (number?, ...any)
+		local entityId, entityRecord = next(world.entities, lastEntityId)
+		if entityId == nil or entityRecord == nil then
+			return nil
+		end
+
+		lastEntityId = entityId
+
+		local componentIdToComponent = world.componentIdToComponent
+		local archetype = entityRecord.archetype
+		local componentInstances = {}
+		for index, componentStorage in archetype.storage do
+			componentInstances[componentIdToComponent[archetype.storageIndexToComponentId[index]]] =
+				componentStorage[entityRecord.indexInArchetype]
+		end
+
+		return entityId, componentInstances
+	end
+end
+
+local function ensureArchetype(world: World, componentIds: { number })
+	local archetypeId = archetypeOf(componentIds)
+	local archetypes = world.archetypes
+	for _, archetype in archetypes do
```

Comment:
Linear traversal to lookup archetypes is really slow. Consider when you have thousands of archetypes.